crates/core/component/auction/src/component/dutch_auction.rs

```diff
@@ -51,11 +64,183 @@ pub(crate) trait DutchAuctionManager: StateWrite {
+        } else {
+            Balance::zero()
+        };
+
+        // We remove the execution trigger that we are currently processing:
+        self.unset_trigger_for_dutch_id(auction_id, trigger_height);
+
+        // Prepare a new auction, based on the previous one.
+        let mut new_dutch_auction = DutchAuction {
+            description: old_dutch_auction.description,
+            state: old_dutch_auction.state,
+        };
+
+        let balance_input_asset = lp_reserves
+            .provided()
+            .filter(|v| v.asset_id == input.asset_id)
+            .map(|v| v.amount)
+            .sum::<Amount>();
+        let balance_output_asset = lp_reserves
+            .provided()
+            .filter(|v| v.asset_id == output_id)
+            .map(|v| v.amount)
+            .sum::<Amount>();
+
+        new_dutch_auction.state.current_position = None;
+        // We overwrite the input amount
+        new_dutch_auction.state.input_reserves = balance_input_asset;
+        new_dutch_auction.state.output_reserves += balance_output_asset;
+        new_dutch_auction.state.next_trigger = 0;
+
+        // Compute the current step index, between 0 and `step_count`.
+        let step_index = auction_trigger
+            .compute_step_index(trigger_height)
+            .expect("trigger data is validated");
+
+        if step_index >= step_count {
+            // If the termination condition has been reached, we set the auction
+            // sequence to 1 (Closed).
+            new_dutch_auction.state.sequence = 1;
+        } else {
+            // Otherwise, we compute the next trigger height and generate a liquidity
+            // position for the new auction round.
+            let next_trigger = auction_trigger
+                .compute_next_trigger_height(trigger_height)
+                .expect("trigger data is validated")
+                .expect("the step count has not been reached");
+
+            // We compute the price parameters for the LP:
+            let (p, q) = compute_pq_at_step(&new_dutch_auction.description, step_index);
+
+            // Next, we want to construct an auction LP position and send it to the DEX.
+            // The nonce must be chosen so that the resulting position id is unique:
+            // `PositionManager::open_position` will reject duplicates.
+            //
+            // To do this, we keep track of our number of attempts at opening a position,
+            // and compute:
+            // position_nonce = H(auction_nonce || step_index || attempt_counter)
+            // until the resulting position id (based on the nonce) is unique and accepted
+            // by the DEX.
+            let mut attempt_counter = 0u64;
+            loop {
+                let pair = DirectedTradingPair::new(auction_input_id, auction_output_id);
+                let lp_reserves = Reserves {
+                    r1: new_dutch_auction.state.input_reserves,
+                    r2: Amount::zero(),
+                };
```

Comment:
This seems to be double-counting the tokens. Either they're in the LP position (in which case they should not be in the auction) or they're in the auction object (in which they should not be in the LP reserves).

Reply:
Oh yeah indeed, this is wrong

Reply:
xref https://github.com/penumbra-zone/penumbra/pull/4254/files#r1575464539

I think that the input reserves of the auction should be zero if the tokens are in an opened LP position

Reply:
So, the auction state has a bunch of reserves $\Delta, \Lambda$, at each price point $p_1, p_2... p_i$ it auctions off a portion of that input, when we roll out the position (close -> open) we deduct the portion that has been sold (`balance_input_asset`) and accumulate what the LP has been able to trade:

```rust
        new_dutch_auction.state.input_reserves = balance_input_asset;
        new_dutch_auction.state.output_reserves += balance_output_asset;
```

By the moment we build those reserves, we should have closed and withdrawn the LP and credited the auction state with the balance from that position, what am I missing?

Reply:
Erratum, it auctions *all* of that input but only a portion gets sold. The unfilled part gets returned.

Reply:
The auction object can own tokens in two ways: directly, in its own structure, or indirectly, via the LP position.

On line 118, value is moved out of the LP position (if it exists) and onto the stack (in `lp_reserves`, destructured to `balance_input_asset` and `balance_output_asset`.

On line 168, `balance_output_asset` is consumed by accumulating it into `state.output_reserves`. I agree this is correct.

On line 167, any previous value of `state.input_reserves` is burned, and `balance_input_asset` is consumed by overwriting it. Then on line 204, the value is duplicated into the position if one is created. This would be an inflation, except that any future pass through _this method only_ would happen to burn one of the copies when it re-executes line 167. However, this is not the only possibility -- if the auction is closed, that runs a different codepath, which accumulates the position reserves together with the auction's reserves, on line 313. I believe this would trigger an inflation bug.

Instead, we should avoid this problem by ensuring that we don't duplicate value. The auction's `input_reserves` should be zeroed out when the value moves into an LP position and restored when value is returned.

Reply:
Got it, makes sense, thanks